ra2ce/graph/osm_network_wrapper.py

```diff
@@ -0,0 +1,266 @@
+                                 graph: MultiDiGraph):
+        """
+        checks if both extremities are in the unique_graph (u has not the same coor of v, no line from u to itself is
+        allowed). Checks if an edge is already made between such extremities with the given id and coordinates before
+        considering it in the unique graph
+        """
+        if unique_elements is None or len(unique_elements) == 0:
+            raise ValueError("""unique_elements cannot be None or empty. 
+            Provide a set with all unique node coordinates as tuples of (x, y)""")
+        if unique_graph is None:
+            raise ValueError("""unique_graph cannot be None. Provide a graph with unique nodes or perform the 
+        drop_duplicates_in_nodes on the graph to generate a unique_graph""")
+        for u, v, data in graph.edges(data=True):
+            extremities_data = ExtremitiesData.get_extremities_data_for_sub_graph(from_node_id=u, to_node_id=v,
+                                                                                  sub_graph=unique_graph,
+                                                                                  graph=graph,
+                                                                                  shared_elements=unique_elements)
+            if extremities_data.from_id != extremities_data.to_id:  # excluding a self-directed link
+                if extremities_data.from_to_id is not None and extremities_data.from_to_coor is not None:
+                    if (extremities_data.from_to_id, extremities_data.to_from_id) not in unique_elements and \
+                            (extremities_data.from_to_coor, extremities_data.to_from_coor) not in unique_elements:
+                        edge_attributes = {
+                            key: value for key, value in data.items()
+                        }
+                        unique_graph.add_edge(extremities_data.from_id, extremities_data.to_id, **edge_attributes)
+                        unique_elements.add((extremities_data.from_to_id, extremities_data.to_from_id))
+                        unique_elements.add((extremities_data.from_to_coor, extremities_data.to_from_coor))
+
+        return unique_graph
+
+    @staticmethod
+    def find_node_id_by_coor(graph: MultiDiGraph, target_x: float, target_y: float):
+        """
+         finds the node in unique graph with the same coor
+        """
+        for node, data in graph.nodes(data=True):
+            if 'x' in data and 'y' in data and data['x'] == target_x and data['y'] == target_y:
+                return node
+        return None
+
+    @staticmethod
+    def snap_nodes_to_nodes(graph, threshold):
+        return consolidate_intersections(G=graph, rebuild_graph=True,
+                                         tolerance=threshold, dead_ends=False)
+
+    @staticmethod
+    def snap_nodes_to_edges(graph, threshold):
+        pass
+
+
+@dataclass
+class ExtremitiesData:
```

Comment:
Extract this into a separate class.
You can, and should, extract both this and the `osm_network_wrapper` into their own module.